Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -17,6 +17,112 @@ import NIO
 let crlf: StaticString = "\r\n"
 let headerSeparator: StaticString = ": "
 
+fileprivate let comma = UInt8(",".utf8CString[0])
+fileprivate let whiteSpaces = [" ", "\t"]
+                              .map( {UInt8($0.utf8CString[0])} )
+
+/// An `IteratorProtocol` that can iterate through comma separated list of values for a certain
+/// header
+///
+///  **Example:**
+///
+/// Suppose you have this headers:
+///
+///      Connection: keep-alive, x-server
+///      Content-Type: text/html
+///      Connection: other
+///
+/// You can iterate using this struct on those headers, for values of `Connection`, to get
+/// `keep-alive`, then `x-server`, then `other`
+///
+public struct HTTPListHeaderIterator<T: Collection>: IteratorProtocol where T.Element == UInt8 {
+    
+    public typealias Element = ByteBufferView
+    
+    private var isNowOnAField = false
+    private var currentHeaderIndex: Int = -1
+    private var byteBuffer: ByteBuffer
+    private var singleValueViewIterator: Array<ByteBufferView>.Iterator
+    private let headerName: T
+    private let headers: [HTTPHeader]
+    
+    /// Returns next index in headers
+    ///
+    /// - Parameter current: The index to begin iteration at
+    /// - Returns: The next index of the header in header array, or `nil` if not found
+    internal func nextHeaderIndex(ifCurrentIs current: Int) -> Int? {
+        for (idx, currentHeader) in headers.enumerated().dropFirst(current + 1) {
+            if byteBuffer.getSlice(at: currentHeader.name.start,
+                                   length: currentHeader.name.length)?.readableBytesView
+                .compareReadableBytes(to: headerName) ?? false {
+                return idx
+            }
+        }
+        return nil
+    }
+
+    mutating public func next() -> ByteBufferView? {
+        // Still not reading a specific field
+        if !isNowOnAField {
+            // Try to get next match
+            guard let index = nextHeaderIndex(ifCurrentIs: currentHeaderIndex) else { return nil }
+            // If succeeded, prepare for parsing it
+            currentHeaderIndex = index
+            isNowOnAField = true
+            singleValueViewIterator =
+                byteBuffer.viewBytes(at: headers[currentHeaderIndex].value.start,
+                                     length: headers[currentHeaderIndex].value.length)
+                    .split(separator: comma).makeIterator()
+        }
+        
+        let next = singleValueViewIterator.next()
+        // Null or empty buffer?
+        if (next?.count ?? 0) == 0 {
+            // Try again
+            isNowOnAField = false
+            return self.next()
+        }
+        return next?.trimSpaces
```

Comment:
The structure of this method is a bit odd, because it forces us to save a bit of global state we don't need. Instead, I think we can resolve this by making `singleValueViewIterator` optional, and rewriting to this:

```swift
if let next = self.singleValueViewIterator?.next() {
    return next.trimSpaces
} else {
    // End of this buffer. Let's try to grab the next one.
    guard let index = self.headerIndex(after: currentHeaderIndex) else {
        // No more buffers left.
        return nil
    }
    self.currentHeaderIndex = index
    self.singleValueViewIterator = byteBuffer.viewBytes(at: headers[currentHeaderIndex].value.start,
                                                        length: headers[currentHeaderIndex].value.length)
        .split(separator: comma)
        .makeIterator()
    return self.next()
}
```

This makes the structure of the method a lot clearer, and keeps the recursion in a sensible location.